pkg/disttask/framework/taskexecutor/manager.go

```diff
@@ -218,7 +218,9 @@ func (m *Manager) onRunnableTasks(tasks []*proto.Task) {
 		}
 		logutil.Logger(m.logCtx).Info("detect new subtask", zap.Int64("task-id", task.ID))
 
-		if !m.slotManager.canAlloc(task) {
+		canAlloc, tasksNeedFree := m.slotManager.canAlloc(task)
+		m.onCanceledTasks(context.Background(), tasksNeedFree)
```

Comment:
as discussed in group, this method is used to make the task as cancelled, not cancel the task-executor

Reply:
also, we only need to cancel the tasks that has lower order and occupies the slots we need to run current task, not cancel all.

Reply:
cancel(nil) can cancel the running task without modifying the task state.

Reply:
fixed in https://github.com/pingcap/tidb/pull/49713/commits/202ea217f1a8ce8e6ac760a14b36c0ef7a5d4b3d